lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -823,6 +835,44 @@ data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
     , newPassphrase :: !(ApiT (Passphrase "raw"))
     } deriving (Eq, Generic, Show)
 
+-- | Polymorphic wrapper for byte arrays, parameterised by the desired string
+-- encoding.
+newtype ApiBytesT (base :: Base) bs = ApiBytesT { getApiBytesT :: bs }
+    deriving (Generic, Show, Eq, Functor)
+    deriving newtype (Semigroup, Monoid, Hashable)
+    deriving anyclass NFData
+
+data ApiSerialisedTransaction
+    = ApiSerialisedTransaction (ApiBytesT 'Base64 SerialisedTx)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiTxInputExtended (n :: NetworkDiscriminant) = ApiTxInputExtended
+    { source :: !(Maybe (AddressAmount (ApiT Address, Proxy n)))
+    , input :: !(ApiT TxIn)
+    , derivationPath :: !(NonEmpty (ApiT DerivationIndex))
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+data ApiConstructTransaction (n :: NetworkDiscriminant) = ApiConstructTransaction
+    { serializedTransaction :: !ApiSerialisedTransaction
+    , inputs :: [ApiTxInputExtended n]
+    } deriving (Eq, Generic, Show)
+      deriving anyclass NFData
+
+data ApiMultiDelegationAction = Joining (ApiT PoolId) Natural | Leaving (ApiT PoolId)
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiConstructTransactionData (n :: NetworkDiscriminant) = ApiConstructTransactionData
+    { payments :: !(Maybe (NonEmpty (AddressAmount (ApiT Address, Proxy n))))
+    , withdrawal :: !(Maybe ApiWithdrawalPostData)
+    , metadata :: !(Maybe (ApiT TxMetadata))
+    , mint :: !(Maybe (ApiT W.TokenMap))
+    , delegation :: ![ApiMultiDelegationAction]
```

Comment:
@Anviking does it make sense? see `ApiMultiDelegationAction` - Natural will stand for soft index of stake key. Thanks to that we could in one tx realize several join/leave pool actions. Maybe it could help with portfolio recreation, etc.

Reply:
I think `[ApiMultiDelegationAction]` as a raw/manual/low-level kind of multi-delegation api would be really nice. `Joining (ApiT PoolId) Natural` makes sense to me.

But instead of `Leaving (ApiT PoolId)`, we should presumably have `Leaving Natural`. We don't have to specify a pool id when we de-register a stake key, but we do need to know which stake key we want to deregister.

And I think we currently use the terminology `Join` and `Quit`, so why not stick to it?

Reply:
(I just realised I didn't plan for the `DelegationState` to validate (several) user-provided multi-delegation actions, but it would be nice to support, and I don't see how it could be problematic.)

But it does mean that, as said, the delegation actions need to be validated by the `DelegationState`.

Also I guess you would have to start with only supporting key `0` or maybe even a single action for now, when implementing it.

Reply:
thanks for the look. Now I have : 
```
data ApiMultiDelegationAction = Joining (ApiT PoolId) Natural | Leaving Natural
```